### PR TITLE
INTPYTHON-994 Make QuerySet.exclude() omit documents with absent JSONField keys

### PR DESCRIPTION
**Pull request by timgraham**
`exclude(value__key=X)` and negated Q objects on `JSONField` key lookups wrongly returned documents that don't contain the key. SQL databases treat the comparison on a missing key as `NULL` (three-valued logic), so `NOT(...)` excludes those rows; MongoDB's `$nor`/`$not` instead treat a missing key as "not equal" and included them.

Give `KeyTransformExact` a custom negation (key exists AND value != RHS), and push negation down to the leaves via De Morgan's laws in `where_node()` when a subtree contains such a lookup, so each lookup can supply its own SQL-correct negation while other lookups keep using `$nor`/`$not`.

### CODE REVIEW

File: django_mongodb_backend/query.py [hunk 1]
```diff
@@ -355,7 +355,45 @@ def _get_reroot_replacements(expression):
...
+            return negate(compiler, connection, as_expr=as_expr)
+        mql = node.as_mql(compiler, connection, as_expr=as_expr)
+        return {"$not": [mql]} if as_expr else {"$nor": [mql]}
+    if node.negated:
+        # NOT(NOT(group)) == group; compile the group without negation.
+        node = node.__class__(node.children, node.connector)
+        return node.as_mql(compiler, connection, as_expr=as_expr)
+    operator = "$or" if node.connector == AND else "$and"
+    parts = [_negate(child, compiler, connection, as_expr) for child in node.children]
+    return parts[0] if len(parts) == 1 else {operator: parts}
```
Comment (NoahStapp) [inline django_mongodb_backend/query.py:413]: What if `parts` is empty? Should that give us no rows or all rows?

File: django_mongodb_backend/fields/json.py [hunk 2]
```diff
@@ -228,6 +250,7 @@ def register_json_field():
     KeyTransform.can_use_path = key_transform_is_simple_column
     KeyTransform.is_simple_column = key_transform_is_simple_column
     KeyTransformExact.as_mql_path = key_transform_exact_path
+    KeyTransformExact.as_mql_negated = key_transform_exact_negated
```
Comment (NoahStapp) [inline django_mongodb_backend/fields/json.py]: This only covers `KeyTransformExact`: do other JSON lookups (gt, lt, gte, lte, in, regex) not have the same issue?

**Review by NoahStapp — DISMISSED** [review]

File: django_mongodb_backend/query.py (hunk 1 above)
Comment (timgraham) [reply django_mongodb_backend/query.py:413]: I believe it won't happen:

1. Entry from `where_node()` (line 396): guarded by `_has_custom_negation(self)`, which is `any(... for child in node.children)` — that returns False for a childless node, so we never enter `_negate()` with an empty top node. The entry node always has ≥1 child.
  2. Recursive descent (line 388 iterating into a non-negated child `WhereNode`): Django prunes empty `Q(`)/`WhereNode` objects during query construction. For example, `exclude(Q(value__c__gt=10) & Q())` doesn't produce an empty child node; the empty `Q()` is dropped and the tree is just the lookup plus the `IsNull` term Django adds for negated key lookups.

File: django_mongodb_backend/fields/json.py (hunk 2 above)
Comment (timgraham) [reply django_mongodb_backend/fields/json.py]: They do. Fixed and added `test_lookup_exclude_other_lookups`. https://github.com/django/django/pull/21567

File: django_mongodb_backend/query.py (hunk 1 above)
Comment (NoahStapp) [reply django_mongodb_backend/query.py:413]: So something like `Document.objects.exclude(value__k="x", num__in=[])` should return everything? For me it returns nothing.

Comment (timgraham) [reply django_mongodb_backend/query.py:413]: You are correct. I fixed this and added a second commit to https://github.com/django/django/pull/21567 to cover the new code: "Added JSONKey key exclude() tests for Empty/FullResultSet cases."

Comment (NoahStapp) [reply django_mongodb_backend/query.py:413]: I still see no documents running `Document.objects.exclude(value__k__in=[])`.

File: django_mongodb_backend/query.py [hunk 3]
```diff
@@ -361,7 +361,63 @@ def _get_reroot_replacements(expression):
...
+        except EmptyResultSet as exc:
+            raise FullResultSet from exc
+        except FullResultSet as exc:
+            raise EmptyResultSet from exc
+        return {"$not": [mql]} if as_expr else {"$nor": [mql]}
+    if node.negated:
+        # NOT(NOT(group)) == group; compile the group without negation.
+        node = node.__class__(node.children, node.connector)
+        return node.as_mql(compiler, connection, as_expr=as_expr)
+    operator = "$or" if node.connector == AND else "$and"
```
Comment (NoahStapp) [inline django_mongodb_backend/query.py:398]: Is it possible for a nested XOR to hit this and incorrectly negate?

**Review by NoahStapp — CHANGES_REQUESTED** [review]

**Comment by timgraham** [comment]
I'm out of time with MongoDB, so I'l have to hand this off.
